<hash_map>: Consistently use `int = 0` SFINAE

Over the years, we settled on a consistent pattern for SFINAE: default template arguments of the form `enable_if_t<CONDITION, int> = 0`. This works in a wide variety of situations, including when we have an overload set where SFINAE makes exactly one overload viable. There are a few remaining occurrences of SFINAE that don't follow this pattern.

In `<hash_map>`, these occurrences are:

https://github.com/microsoft/STL/blob/28ec9a32952e0d7443936f8d5ae5d675ba6cf65c/stl/inc/hash_map#L193
https://github.com/microsoft/STL/blob/28ec9a32952e0d7443936f8d5ae5d675ba6cf65c/stl/inc/hash_map#L198
https://github.com/microsoft/STL/blob/28ec9a32952e0d7443936f8d5ae5d675ba6cf65c/stl/inc/hash_map#L376
https://github.com/microsoft/STL/blob/28ec9a32952e0d7443936f8d5ae5d675ba6cf65c/stl/inc/hash_map#L381

This change is currently blocked by a compiler bug. We need to provide a reduced test case to the compiler front-end team.

(Split from #187.)

Reported the compiler bug as DevCom-848104 "MSVC incorrectly rejects `enable_if_t<Condition, int> = 0` SFINAE":

```
C:\Temp\HASH>type hash.cpp
#include <type_traits>
#include <utility>
using namespace std;

template <typename Traits>
struct HashBase {
    using value_type = typename Traits::value_type;

    void insert(value_type&&) = delete;
};

template <typename K, typename V>
struct HashTraits {
    using value_type = pair<const K, V>;
};

template <typename K, typename V>
struct HashMap : HashBase<HashTraits<K, V>> {
    using MyBase = HashBase<HashTraits<K, V>>;

#ifdef WORKAROUND1
    using value_type = pair<const K, V>;
#else
    using value_type = typename MyBase::value_type;
#endif

    using MyBase::insert;

#ifdef WORKAROUND2
    template <typename T, class = enable_if_t<is_constructible_v<value_type, T>, int>>
#else
    template <typename T, enable_if_t<is_constructible_v<value_type, T>, int> = 0>
#endif

    void insert(T&&) {}
};

struct Movable {
    Movable() = default;
    Movable(Movable&&) = default;
    Movable& operator=(Movable&&) = default;

    Movable(const Movable&) = delete;
    Movable& operator=(const Movable&) = delete;
};

int main() {
    HashMap<Movable, int> hm;
    pair<Movable, int> p(Movable{}, 3);
    hm.insert(move(p));
}

C:\Temp\HASH>cl /EHsc /nologo /W4 /c hash.cpp
hash.cpp
hash.cpp(50): error C2280: 'void HashBase<HashTraits<K,V>>::insert(std::pair<const K,V> &&)': attempting to reference a deleted function
        with
        [
            K=Movable,
            V=int
        ]
hash.cpp(9): note: see declaration of 'HashBase<HashTraits<K,V>>::insert'
        with
        [
            K=Movable,
            V=int
        ]
hash.cpp(9): note: 'void HashBase<HashTraits<K,V>>::insert(std::pair<const K,V> &&)': function was explicitly deleted
        with
        [
            K=Movable,
            V=int
        ]

C:\Temp\HASH>cl /EHsc /nologo /W4 /c /DWORKAROUND1 hash.cpp
hash.cpp

C:\Temp\HASH>cl /EHsc /nologo /W4 /c /DWORKAROUND2 hash.cpp
hash.cpp

C:\Temp\HASH>clang-cl -m32 /EHsc /nologo /W4 /c hash.cpp

C:\Temp\HASH>clang-cl -m32 /EHsc /nologo /W4 /c /DWORKAROUND1 hash.cpp

C:\Temp\HASH>clang-cl -m32 /EHsc /nologo /W4 /c /DWORKAROUND2 hash.cpp

C:\Temp\HASH>
```

--- a/stl/inc/hash_map
+++ b/stl/inc/hash_map
@@ -190,12 +190,12 @@ namespace stdext {
 
         using _Mybase::insert;
 
-        template <class _Valty, class = enable_if_t<is_constructible_v<value_type, _Valty>>> // TRANSITION, GH-249
+        template <class _Valty,  enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
         pair<iterator, bool> insert(_Valty&& _Val) {
             return this->emplace(_STD forward<_Valty>(_Val));
         }
 
-        template <class _Valty, class = enable_if_t<is_constructible_v<value_type, _Valty>>> // TRANSITION, GH-249
+        template <class _Valty,  enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
         iterator insert(const_iterator _Where, _Valty&& _Val) {
             return this->emplace_hint(_Where, _STD forward<_Valty>(_Val));
         }
@@ -373,12 +373,12 @@ namespace stdext {
 
         using _Mybase::insert;
 
-        template <class _Valty, class = enable_if_t<is_constructible_v<value_type, _Valty>>> // TRANSITION, GH-249
+        template <class _Valty,  enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
         iterator insert(_Valty&& _Val) {
             return this->emplace(_STD forward<_Valty>(_Val));
         }
 
-        template <class _Valty, class = enable_if_t<is_constructible_v<value_type, _Valty>>> // TRANSITION, GH-249
+        template <class _Valty,  enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
         iterator insert(const_iterator _Where, _Valty&& _Val) {
             return this->emplace_hint(_Where, _STD forward<_Valty>(_Val));
         }

--- a/stl/inc/hash_map
+++ b/stl/inc/hash_map
@@ -121,7 +121,7 @@ namespace stdext {
         using referent_type   = _Ty;
         using key_compare     = _Tr;
         using value_compare   = typename _Mybase::_Value_compare;
-        using value_type      = typename _Mybase::value_type;
+        using value_type      = pair<const _Kty, _Ty>;
         using allocator_type  = typename _Mybase::allocator_type;
         using size_type       = typename _Mybase::size_type;
         using difference_type = typename _Mybase::difference_type;
@@ -311,7 +311,7 @@ namespace stdext {
         using referent_type   = _Ty; // old name, magically gone
         using key_compare     = _Tr;
         using value_compare   = typename _Mybase::_Value_compare;
-        using value_type      = typename _Mybase::value_type;
+        using value_type      = pair<const _Kty, _Ty>;
         using allocator_type  = typename _Mybase::allocator_type;
         using size_type       = typename _Mybase::size_type;
         using difference_type = typename _Mybase::difference_type;

--- a/stl/inc/hash_map
+++ b/stl/inc/hash_map
@@ -190,12 +190,12 @@ namespace stdext {
 
         using _Mybase::insert;
 
-        template <class _Valty, enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
+        template <class _Valty, enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0>
         pair<iterator, bool> insert(_Valty&& _Val) {
             return this->emplace(_STD forward<_Valty>(_Val));
         }
 
-        template <class _Valty, enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
+        template <class _Valty, enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0>
         iterator insert(const_iterator _Where, _Valty&& _Val) {
             return this->emplace_hint(_Where, _STD forward<_Valty>(_Val));
         }
@@ -373,12 +373,12 @@ namespace stdext {
 
         using _Mybase::insert;
 
-        template <class _Valty, enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
+        template <class _Valty, enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0>
         iterator insert(_Valty&& _Val) {
             return this->emplace(_STD forward<_Valty>(_Val));
         }
 
-        template <class _Valty, enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
+        template <class _Valty, enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0>
         iterator insert(const_iterator _Where, _Valty&& _Val) {
             return this->emplace_hint(_Where, _STD forward<_Valty>(_Val));
         }

--- a/stl/inc/hash_map
+++ b/stl/inc/hash_map
@@ -190,12 +190,12 @@ namespace stdext {
 
         using _Mybase::insert;
 
-        template <class _Valty,  enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
+        template <class _Valty, enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
         pair<iterator, bool> insert(_Valty&& _Val) {
             return this->emplace(_STD forward<_Valty>(_Val));
         }
 
-        template <class _Valty,  enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
+        template <class _Valty, enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
         iterator insert(const_iterator _Where, _Valty&& _Val) {
             return this->emplace_hint(_Where, _STD forward<_Valty>(_Val));
         }
@@ -373,12 +373,12 @@ namespace stdext {
 
         using _Mybase::insert;
 
-        template <class _Valty,  enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
+        template <class _Valty, enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
         iterator insert(_Valty&& _Val) {
             return this->emplace(_STD forward<_Valty>(_Val));
         }
 
-        template <class _Valty,  enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
+        template <class _Valty, enable_if_t<is_constructible_v<value_type, _Valty>, int> = 0> // TRANSITION, GH-249
         iterator insert(const_iterator _Where, _Valty&& _Val) {
             return this->emplace_hint(_Where, _STD forward<_Valty>(_Val));
         }

